src/service/url-expander/parser.js

```diff
@@ -0,0 +1,227 @@
+      rethrowAsync(e);
+      return Promise.resolve('');
+    }
+  }
+
+  /**
+   * take the template url and return a promise of its evaluated value
+   * @param {string} url url to be substituted
+   * @param {!Object<string, *>=} opt_bindings additional one-off bindings
+   * @param {!Object<string, boolean>=} opt_whiteList Optional white list of names
+   *     that can be substituted.
+   * @return {!Promise<string>}
+   */
+  expand(url, opt_bindings, opt_whiteList) {
+    if (!url.length) {
+      return Promise.resolve(url);
+    }
+    const expr = this.variableSource_.getExpr(opt_bindings, /*opt_ignoreArgs */ true);
+    const matches = this.findMatches_(url, expr);
+    // if no keywords move on
+    if (!matches.length) {
+      return Promise.resolve(url);
+    }
+    const mergedPositions = this.eliminateOverlaps_(matches, url, opt_whiteList);
+    return this.parseUrlRecursively_(url, mergedPositions, opt_bindings);
+  }
+
+  /**
+   * @param {string} url
+   * @param {!Array<Object<string, string|number>>} matches array of objects representing
+   *  matching keywords
+   * @param {!Object<string, *>=} opt_bindings additional one-off bindings
+   * @return {!Promise<string>}
+   */
+  parseUrlRecursively_(url, matches, opt_bindings) {
+    const stack = [];
+    let urlIndex = 0;
+    let matchIndex = 0;
+    let match = matches[matchIndex];
+
+    const evaluateNextLevel = () => {
+      let builder = '';
+      const results = [];
+
+      while (urlIndex < url.length && matchIndex <= matches.length) {
+
+        if (match && urlIndex === match.start) {
+          let binding;
+          // find out where this keyword is coming from
+          if (opt_bindings && opt_bindings.hasOwnProperty(match.name)) {
+            // the optional bindings
+            binding = opt_bindings[match.name];
+          } else {
+            // or the global source
+            binding = this.variableSource_.get(match.name);
+            binding = binding.async || binding.sync;
+          }
+
+          urlIndex = match.stop + 1;
+          match = matches[++matchIndex];
+
+          if (url[urlIndex] === '(') {
+            // if we hit a left parenthesis we still need to get args
+            urlIndex++;
+            stack.push(binding);
+            if (builder.length) {
+              results.push(builder);
+            }
+            results.push(evaluateNextLevel());
+          } else {
+            if (builder.length) {
+              results.push(builder);
+            }
+            results.push(this.evaluateBinding_(binding));
+          }
+
+          builder = '';
+        }
+
+        else if (url[urlIndex] === ',') {
+          if (builder.length) {
+            results.push(builder.trim());
+          }
+          builder = '';
+          urlIndex++;
+
+        }
+
+        else if (url[urlIndex] === ')') {
+          urlIndex++;
+          const binding = stack.pop();
+          const args = [...results, builder.trim()];
```

Comment:
Right now we are trimming all args to support `arg1, arg2` but we may choose to drop this if it increases complexity